### PR TITLE
Global memcap v1

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2316

Describe changes:
This new feature add a new `global-memcap` which is used to rule all the other memcaps available.
If enabled, the sum of all the memcaps cannot be greater than the value set for the global memcap, and in case a memcap is set to zero, it won't be considered for the sum.
This brings a patch also that permits to set the memcaps value for dns and replace the static list of memcap with a dynamic one.
A user could easily see a list of memcaps running suricata with the `--display-memcaps` option:
`suricata --display-memcaps`
The idea of having a list of memcaps can be useful in case of tuning.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/185
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/49

### CODE REVIEW

File: doc/userguide/configuration/suricata-yaml.rst [hunk 1]
```diff
@@ -206,6 +206,20 @@ statement it will overwrite the included file. Therefor any include
 statement at the end of the document will overwrite the already
 configured sections.
 
+Global Memcap
+-------------
+
+It represents the maximum amount of memory that can be used
+to allocate all the memcaps values.
```
Comment (norg) [inline doc/userguide/configuration/suricata-yaml.rst]: memcap without *s*? :)

File: doc/userguide/configuration/suricata-yaml.rst [hunk 2]
```diff
@@ -206,6 +206,20 @@ statement it will overwrite the included file. Therefor any include
...
+The sum of all of them (stream.memcap, flow.memcap, etc.) cannot be
+greater than the global memcap
```
Comment (norg) [inline doc/userguide/configuration/suricata-yaml.rst]: missing dot at end of line.

File: doc/userguide/configuration/suricata-yaml.rst [hunk 3]
```diff
@@ -206,6 +206,20 @@ statement it will overwrite the included file. Therefor any include
...
+
+It can be enabled in suricata.yaml and update through unix-socket
+using the memcap-set command.
```
Comment (norg) [inline doc/userguide/configuration/suricata-yaml.rst]: *updated*

File: suricata.yaml.in [hunk 4]
```diff
@@ -1054,6 +1054,13 @@ asn1-max-frames: 256
 # umask passed on by the shell.
 #umask: 022
 
+# Global memcap
+# This value represents the maximum amount of memory that can be used
+# to allocate all the memcaps values.
```
Comment (norg) [inline suricata.yaml.in]: same as above, maybe memcap values instead of memcaps values

**Comment by victorjulien** [comment]
commit "util-memcap: add feature" needs more explanation

File: src/host.c [hunk 5]
```diff
@@ -202,6 +206,11 @@ void HostInitConfig(char quiet)
             SC_ATOMIC_SET(host_config.memcap, host_memcap);
         }
     }
+    if (GlobalMemcapReached(SC_ATOMIC_GET(host_config.memcap))) {
```
Comment (victorjulien) [inline src/host.c]: could you create a unit func for this? Right now we see this same code block many times. E.g.:
```c
    GlobalMemcapSomename(SC_ATOMIC_GET(host_config.memcap), "host.memcap");
```
Error handling and 'exit' could then be handled in the new util func.

File: src/app-layer-htp-mem.c [hunk 6]
```diff
@@ -58,9 +62,18 @@ void HTPParseMemcap()
                        conf_val);
             exit(EXIT_FAILURE);
         } else {
-            SC_ATOMIC_SET(htp_config_memcap, memcap);
+            if (GlobalMemcapReached(memcap)) {
```
Comment (victorjulien) [inline src/app-layer-htp-mem.c]: same here: this looks a lot like stream, ftp and maybe other places. A util would be much nicer.

**Pull request closed without merging**